### PR TITLE
HDFS cat support

### PR DESCRIPTION
**Pull request by tammam1998**
Extends the distributed execution mode to handle HDFS cat commands. The PR contributes changes on two levels: 
Compiler: 
- To handle HDFS I introduced a new a`RemoteFileResource` interface and `HDFSResource` implements this interface. The hdfs file resource contains information about a block and all the machines the block exists on.
- HDFSCat DFGNode is added. This node is then removed in the graph optimizing step and replaced by Cat nodes where each cat fid has an HDFSResource that corresponds to a block of the HDFS file.

Distributed Execution:  The worker manager policy now takes into account data availability when choosing a worker.

### CODE REVIEW

File: compiler/definitions/ir/file_id.py [hunk 1]
```diff
@@ -118,3 +121,13 @@ def isNull(self):
...
     def get_ident(self):
         return self.ident
+
+    def is_available_on(self, host):
+        if self.is_ephemeral():
+            return True
+        elif self.has_remote_file_resource():
+            return self.resource.is_available_on(host)
+        else:
+            # Currently any other resource types should
```
Comment (angelhof) [inline compiler/definitions/ir/file_id.py:131]: And if I understand correctly, the main graph never calls `is_available_on`?

File: compiler/definitions/ir/resource.py [hunk 2]
```diff
@@ -56,3 +57,49 @@ def __eq__(self, other):
...
+        host = addr.rsplit(":", 1)[0]
+        normalized_host = socket.gethostbyaddr(host)[2][0]
+        return normalized_host
+
+class HDFSFileResource(RemoteFileResource):
+    ## The uri is the path of the file.
+    def __init__(self, uri, resource_hosts):
+        """
+        Params:
+            uri: Usually the path to the file
```
Comment (angelhof) [inline compiler/definitions/ir/resource.py:86]: Is the path to the file the same in all machines? If not, let's add a TODO here to fix that later.

File: compiler/definitions/ir/resource.py [hunk 3]
```diff
@@ -56,3 +57,49 @@ def __eq__(self, other):
...
+    def __eq__(self, other):
+        if isinstance(other, HDFSFileResource):
+            return self.uri == other.uri
+        return False
+
+    def is_available_on(self, host):
+        return host in self.hosts
+
+    def __repr__(self):
+        return f'hdfs://{str(self.uri)}'
```
Comment (angelhof) [inline compiler/definitions/ir/resource.py:102]: I don't think you need to add `str` inside format strings. For example,  `f'{1}` should work and call __repr__ for the value automatically.

File: compiler/dspash/hdfs_file_data.py [hunk 4]
```diff
@@ -32,51 +32,48 @@ def paths(self):
             )
         return filepaths
 
+    def _getinfo(self, filename):
```
Comment (angelhof) [inline compiler/dspash/hdfs_file_data.py:35]: It is a bit scary (and unexpected) that when calling the initialization of an object we perform a subprocess call :O. Maybe that can happen in a separate function that is clearly called before initialization, in order to make it clear that here we perform a subprocess call, which could fail, etc...

File: compiler/definitions/ir/file_id.py (hunk 1 above)
Comment (tammam1998) [reply compiler/definitions/ir/file_id.py:131]: yeah, this function is for evaluating if a fid can be serviced on the provided host. The "main graph" by design runs in the user's shell so we should never call is_available_on on the main graph fids.

The calls for this function are in worker_manager.py

File: compiler/definitions/ir/resource.py (hunk 2 above)
Comment (tammam1998) [reply compiler/definitions/ir/resource.py:86]: The path is the same for hdfs; the top directory however could be different. That's why the str for HDFSResource returns "$HDFS_DATANODE_DIR/{self.uri}". I set the HDFS_DATANODE_DIR in the worker's environment.

**Review by angelhof — APPROVED** [review]
looks great to me (modulo minor comments)!

File: compiler/dspash/hdfs_file_data.py (hunk 4 above)
Comment (tammam1998) [reply compiler/dspash/hdfs_file_data.py:35]: Ok, I could restructure this to be a function call with the object as a return.

File: compiler/definitions/ir/file_id.py (hunk 1 above)
Comment (angelhof) [reply compiler/definitions/ir/file_id.py:131]: cool!

File: compiler/definitions/ir/resource.py (hunk 2 above)
Comment (angelhof) [reply compiler/definitions/ir/resource.py:86]: Great, a comment on that would be nice :)

File: compiler/dspash/hdfs_file_data.py (hunk 4 above)
Comment (angelhof) [reply compiler/dspash/hdfs_file_data.py:35]: Sounds good

**Pull request merged**